### PR TITLE
feat: networked message channel

### PR DESCRIPTION
**Pull request by LPLafontaineB**
### Description

This PR adds a networked message channel to our DI infrastructure. It uses custom messages to send messages to clients when the server publishes a message through it. This would be used for #601.

### Issue Number(s)

This will be useful for [MTT-2678](https://jira.unity3d.com/browse/MTT-2678) & [MTT-2318](https://jira.unity3d.com/browse/MTT-2318)
### Contribution checklist
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [ ] JIRA ticket ID is in the PR title or at least one commit message
 - [ ] Include the ticket ID number within the body message of the PR to create a hyperlink

### CODE REVIEW

**Comment by LPLafontaineB** [comment]
There is no Jira ticket yet, it is simply something I put together while working on alternative solutions for #601 and I felt it might be interesting as a separate PR to review

**Comment by pdeschain** [comment]
I like this idea a lot.

**Review by pdeschain — APPROVED** [review]

File: Assets/BossRoom/Scripts/Shared/Infrastructure/PubSub/NetworkedMessageChannel.cs [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+    {
+        string m_Name;
+
+        int m_BufferSize;
+
+        bool m_HasRegisteredHandler;
+
+        public NetworkedMessageChannel(string name, int bufferSize)
+        {
+            m_Name = name;
```
Comment (SamuelBellomo) [inline Assets/BossRoom/Scripts/Shared/Infrastructure/PubSub/NetworkedMessageChannel.cs:24]: name could be deduced from type T? I would make sure to append "NetworkedMessageChannel" in from of it to make sure there's no future name conflict

**Review by pdeschain — DISMISSED** [review]

File: Assets/BossRoom/Scripts/Shared/Infrastructure/PubSub/NetworkedMessageChannel.cs [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+                }
+
+                return base.Subscribe(handler);
+            }
+
+            Debug.LogError("Cannot subscribe to NetworkedMessageChannel. NetworkManager is not initialized.");
+            return null;
+        }
+
+        void OnClientDisconnect(ulong clientId)
```
Comment (fernando-cortez) [inline Assets/BossRoom/Scripts/Shared/Infrastructure/PubSub/NetworkedMessageChannel.cs:54]: This gets fired on the server for any client disconnect, so you may need to check if the clientId matches the LocalClientId before unregistering.

Comment (LPLafontaineB) [reply Assets/BossRoom/Scripts/Shared/Infrastructure/PubSub/NetworkedMessageChannel.cs:54]: good point!

Comment (LPLafontaineB) [reply Assets/BossRoom/Scripts/Shared/Infrastructure/PubSub/NetworkedMessageChannel.cs:54]: Another thing with this is that if a subscriber has a lifetime bigger than a networked session, it needs to make sure to unsubscribe when disconnecting and re-subscribe if connecting to a new session. I'll add something to that effect in the class summary

**Review by SamuelBellomo — APPROVED** [review]

**Review by SamuelBellomo — APPROVED** [review]

**Review by SamuelBellomo — APPROVED** [review]

**Pull request merged**